airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py

```diff
@@ -301,58 +301,159 @@ def reset(self) -> None:
+            ).date()
+
+    def _validate_project_timezone(self) -> None:
+        if self.config.get("project_timezone"):
+            try:
+                pendulum.timezone(self.config["project_timezone"])
+            except pendulum.tz.zoneinfo.exceptions.InvalidTimezone as e:
+                raise_config_error(f"Could not parse time zone: {self.config['project_timezone']}, please enter a valid timezone.", e)
+
+    def _validate_end_date(self) -> None:
+        date_str = self.config.get("end_date")
+        try:
+            return pendulum.parse(date_str).date()
+        except pendulum.parsing.exceptions.ParserError as e:
+            raise_config_error(f"time data '{date_str}' does not match format '%Y-%m-%dT%H:%M:%SZ'", e)
 
-        params = {"authenticator": SourceMixpanel.get_authenticator(self.config), "region": self.config.get("region")}
-        project_id = self.config.get("credentials", {}).get("project_id")
-        if project_id:
-            params["project_id"] = project_id
+    def get_request_params(
+        self,
+        *,
+        stream_state: Optional[StreamState] = None,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> MutableMapping[str, Any]:
+        request_params = super().get_request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
 
-        schema["additionalProperties"] = self.config.get("select_properties_by_default", True)
+        start_time = stream_slice.cursor_slice.get("start_time")
 
-        # read existing Engage schema from API
-        schema_properties = EngageSchema(**params).read_records(sync_mode=SyncMode.full_refresh)
-        for property_entry in schema_properties:
-            property_name: str = property_entry["name"]
-            property_type: str = property_entry["type"]
-            if property_name.startswith("$"):
-                # Just remove leading '$' for 'reserved' mixpanel properties name, example:
-                # from API: '$browser'
-                # to stream: 'browser'
-                property_name = property_name[1:]
-            # Do not overwrite 'standard' hard-coded properties, add 'custom' properties
-            if property_name not in schema["properties"]:
-                schema["properties"][property_name] = types.get(property_type, {"type": ["null", "string"]})
-        self.schema = schema
-        return schema
+        from_date_value = (pendulum.parse(start_time) - timedelta(seconds=self._from_date_lookback_window)).date()
+        to_date_value = self._end_date
+        time_value = int((pendulum.parse(start_time) - timedelta(seconds=self._time_lookback_window)).timestamp())
+
+        request_params["from_date"] = from_date_value.format("YYYY-MM-DD")
+        request_params["to_date"] = to_date_value.format("YYYY-MM-DD")
+        request_params["where"] = f'properties["$time"]>=datetime({time_value})'
```

Comment:
Just to make sure I understand the current solution: We don't have slices of 30 days anymore. This is fine because it was slow and the very low rate limiting anyway, we can't be concurrent and there is risk is doing one query per 30 days instead of one query per page

Reply:
defined export as `StateDelegatingStream`, so for the full refresh read we have 30 days slices, which is more safe for the first read.